### DISCUSSION TITLE
Calculation of ArduinoJson buffer size and memory consumption required for AutoConnectElements

### DISCUSSION THREAD

**Discussion by ghmpi** [Hieromon/AutoConnect#595] [Q&A] [answered]
I seem to be having a problem, maybe a bug or a size limit, I'm not too sure.
If put too many large ACSelect elements (pulldown boxes) in a custom page, it will not work. The title on the menubar does not display and the page where the html is supposed to be.. in this example "/alarm".. it gives me a 404 error.

I don't think it's a limit on the actual size of the char itself, as I can make it pretty big with other text and various elements and it will still work. It's just it doesn't seem to like to many, or long ACSelects.

One curious behavior is the SAME code works on a ESP32wrover module, but moving the code to a ESP32-C3 will cause this problem and apparent limit.

I turned on debug on both autoconnect and pagebuilder. The only line about the alarm page is:
[AC] /alarm not registered, handler ignored

Sample page code below, first one works with a limited number of select options in ALARMxMIN, second one that has more pulldown choices does not work on a ESP32-C3.

Thank you!

Regards,
-Moses

//// THIS IS WORKING WORKING

const char PAGE_ALARM[] PROGMEM = R"(
{
  "uri": "/alarm",
  "title": "Alarm",
  "menu": true,
  "element": [
    { "name": "SETTING_ALARM1ENABLE", "type": "ACCheckbox", "value": "1", "label": "Alarm 1 Enable", "checked": "false", "global": "true"},
    { "name": "SETTING_ALARM1SUN", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1MON", "type": "ACCheckbox", "value": "1", "label": "M", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1TUE", "type": "ACCheckbox", "value": "1", "label": "T", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1WED", "type": "ACCheckbox", "value": "1", "label": "W", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1THU", "type": "ACCheckbox", "value": "1", "label": "Th", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1FRI", "type": "ACCheckbox", "value": "1", "label": "F", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1SAT", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false"},
    { "name": "SETTING_ALARM1HOUR", "type": "ACSelect", "option":  [ "00 / 12 AM", "01 / 1 AM", "02 / 2 AM", "03 / 3 AM", "04 / 4 AM", "05 / 5 AM", "06 / 6 AM", "07 / 7 AM", "08 / 8 AM", "09 / 9 AM", "10 / 10 AM", "11 / 11 AM", "12 / 12 PM", "13 / 1 PM", "14 / 2 PM", "15 / 3 PM", "16 / 4 PM", "17 / 5 PM", "18 / 6 PM", "19 / 7 PM", "20 / 8 PM", "21 / 9 PM", "22 / 10 PM", "23 / 11 PM" ], "label": "Time", "posterior": "none"},
    { "name": "SETTING_ALARM1MIN", "type": "ACSelect", "option":  [ "00", "05", "10", "15", "20", "25", "30", "35", "40", "45", "50", "55" ], "label": "", "posterior": "none"},
    { "name": "SETTING_ALARM1DUR", "type": "ACSelect", "option":  [ "Until canceled", "2 seconds", "5 seconds", "10 seconds", "30 seconds", "1 minute", "2 minutes", "3 minutes","4 minutes", "5 minutes", "6 minutes", "7 minutes", "8 minutes", "9 minutes", "10 minutes" ], "label": "Duration"},
    { "name": "newline1", "type": "ACElement", "value": "<hr>"},

    { "name": "SETTING_ALARM2ENABLE", "type": "ACCheckbox", "value": "1", "label": "Alarm 2 Enable", "checked": "false", "global": "true"},
    { "name": "SETTING_ALARM2SUN", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2MON", "type": "ACCheckbox", "value": "1", "label": "M", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2TUE", "type": "ACCheckbox", "value": "1", "label": "T", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2WED", "type": "ACCheckbox", "value": "1", "label": "W", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2THU", "type": "ACCheckbox", "value": "1", "label": "Th", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2FRI", "type": "ACCheckbox", "value": "1", "label": "F", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2SAT", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false"},
    { "name": "SETTING_ALARM2HOUR", "type": "ACSelect", "option":  [ "00 / 12 AM", "01 / 1 AM", "02 / 2 AM", "03 / 3 AM", "04 / 4 AM", "05 / 5 AM", "06 / 6 AM", "07 / 7 AM", "08 / 8 AM", "09 / 9 AM", "10 / 10 AM", "11 / 11 AM", "12 / 12 PM", "13 / 1 PM", "14 / 2 PM", "15 / 3 PM", "16 / 4 PM", "17 / 5 PM", "18 / 6 PM", "19 / 7 PM", "20 / 8 PM", "21 / 9 PM", "22 / 10 PM", "23 / 11 PM" ], "label": "Time", "posterior": "none"},
    { "name": "SETTING_ALARM2MIN", "type": "ACSelect", "option":  [ "00", "05", "10", "15", "20", "25", "30", "35", "40", "45", "50", "55" ], "label": "", "posterior": "none"},
    { "name": "SETTING_ALARM2DUR", "type": "ACSelect", "option":  [ "Until canceled", "2 seconds", "5 seconds", "10 seconds", "30 seconds", "1 minute", "2 minutes", "3 minutes","4 minutes", "5 minutes", "6 minutes", "7 minutes", "8 minutes", "9 minutes", "10 minutes" ], "label": "Duration"},
    { "name": "newline2", "type": "ACElement", "value": "<hr>"},

    { "name": "SETTING_ALARM3ENABLE", "type": "ACCheckbox", "value": "1", "label": "Alarm 3 Enable", "checked": "false", "global": "true"},
    { "name": "SETTING_ALARM3SUN", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3MON", "type": "ACCheckbox", "value": "1", "label": "M", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3TUE", "type": "ACCheckbox", "value": "1", "label": "T", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3WED", "type": "ACCheckbox", "value": "1", "label": "W", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3THU", "type": "ACCheckbox", "value": "1", "label": "Th", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3FRI", "type": "ACCheckbox", "value": "1", "label": "F", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3SAT", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false"},
    { "name": "SETTING_ALARM3HOUR", "type": "ACSelect", "option":  [ "00 / 12 AM", "01 / 1 AM", "02 / 2 AM", "03 / 3 AM", "04 / 4 AM", "05 / 5 AM", "06 / 6 AM", "07 / 7 AM", "08 / 8 AM", "09 / 9 AM", "10 / 10 AM", "11 / 11 AM", "12 / 12 PM", "13 / 1 PM", "14 / 2 PM", "15 / 3 PM", "16 / 4 PM", "17 / 5 PM", "18 / 6 PM", "19 / 7 PM", "20 / 8 PM", "21 / 9 PM", "22 / 10 PM", "23 / 11 PM" ], "label": "Time", "posterior": "none"},
    { "name": "SETTING_ALARM3MIN", "type": "ACSelect", "option":  [ "00", "05", "10", "15", "20", "25", "30", "35", "40", "45", "50", "55" ], "label": "", "posterior": "none"},
    { "name": "SETTING_ALARM3DUR", "type": "ACSelect", "option":  [ "Until canceled", "2 seconds", "5 seconds", "10 seconds", "30 seconds", "1 minute", "2 minutes", "3 minutes","4 minutes", "5 minutes", "6 minutes", "7 minutes", "8 minutes", "9 minutes", "10 minutes" ], "label": "Duration"},
    { "name": "newline3", "type": "ACElement", "value": "<hr>"},

    { "name": "add", "type": "ACSubmit", "value": "Submit", "uri": "/alarm"}
  ]
}
)";

/// THIS IS NOT WORKING, DIFFERENCE IS ADDITIONAL VALUES in SETTING_ALARM1MIN

const char PAGE_ALARM[] PROGMEM = R"(
{
  "uri": "/alarm",
  "title": "Alarm",
  "menu": true,
  "element": [
    { "name": "SETTING_ALARM1ENABLE", "type": "ACCheckbox", "value": "1", "label": "Alarm 1 Enable", "checked": "false", "global": "true"},
    { "name": "SETTING_ALARM1SUN", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1MON", "type": "ACCheckbox", "value": "1", "label": "M", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1TUE", "type": "ACCheckbox", "value": "1", "label": "T", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1WED", "type": "ACCheckbox", "value": "1", "label": "W", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1THU", "type": "ACCheckbox", "value": "1", "label": "Th", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1FRI", "type": "ACCheckbox", "value": "1", "label": "F", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM1SAT", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false"},
    { "name": "SETTING_ALARM1HOUR", "type": "ACSelect", "option":  [ "00 / 12 AM", "01 / 1 AM", "02 / 2 AM", "03 / 3 AM", "04 / 4 AM", "05 / 5 AM", "06 / 6 AM", "07 / 7 AM", "08 / 8 AM", "09 / 9 AM", "10 / 10 AM", "11 / 11 AM", "12 / 12 PM", "13 / 1 PM", "14 / 2 PM", "15 / 3 PM", "16 / 4 PM", "17 / 5 PM", "18 / 6 PM", "19 / 7 PM", "20 / 8 PM", "21 / 9 PM", "22 / 10 PM", "23 / 11 PM" ], "label": "Time", "posterior": "none"},
    { "name": "SETTING_ALARM1MIN", "type": "ACSelect", "option":  [ "00", "01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "40", "41", "42", "43", "44", "45", "46", "47", "48", "49", "50", "51", "52", "53", "54", "55", "56", "57", "58", "59" ], "label": "", "posterior": "none"},
    { "name": "SETTING_ALARM1DUR", "type": "ACSelect", "option":  [ "Until canceled", "2 seconds", "5 seconds", "10 seconds", "30 seconds", "1 minute", "2 minutes", "3 minutes","4 minutes", "5 minutes", "6 minutes", "7 minutes", "8 minutes", "9 minutes", "10 minutes" ], "label": "Duration"},
    { "name": "newline1", "type": "ACElement", "value": "<hr>"},

    { "name": "SETTING_ALARM2ENABLE", "type": "ACCheckbox", "value": "1", "label": "Alarm 2 Enable", "checked": "false", "global": "true"},
    { "name": "SETTING_ALARM2SUN", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2MON", "type": "ACCheckbox", "value": "1", "label": "M", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2TUE", "type": "ACCheckbox", "value": "1", "label": "T", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2WED", "type": "ACCheckbox", "value": "1", "label": "W", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2THU", "type": "ACCheckbox", "value": "1", "label": "Th", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2FRI", "type": "ACCheckbox", "value": "1", "label": "F", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM2SAT", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false"},
    { "name": "SETTING_ALARM2HOUR", "type": "ACSelect", "option":  [ "00 / 12 AM", "01 / 1 AM", "02 / 2 AM", "03 / 3 AM", "04 / 4 AM", "05 / 5 AM", "06 / 6 AM", "07 / 7 AM", "08 / 8 AM", "09 / 9 AM", "10 / 10 AM", "11 / 11 AM", "12 / 12 PM", "13 / 1 PM", "14 / 2 PM", "15 / 3 PM", "16 / 4 PM", "17 / 5 PM", "18 / 6 PM", "19 / 7 PM", "20 / 8 PM", "21 / 9 PM", "22 / 10 PM", "23 / 11 PM" ], "label": "Time", "posterior": "none"},
    { "name": "SETTING_ALARM2MIN", "type": "ACSelect", "option":  [ "00", "05", "10", "15", "20", "25", "30", "35", "40", "45", "50", "55" ], "label": "", "posterior": "none"},
    { "name": "SETTING_ALARM2DUR", "type": "ACSelect", "option":  [ "Until canceled", "2 seconds", "5 seconds", "10 seconds", "30 seconds", "1 minute", "2 minutes", "3 minutes","4 minutes", "5 minutes", "6 minutes", "7 minutes", "8 minutes", "9 minutes", "10 minutes" ], "label": "Duration"},
    { "name": "newline2", "type": "ACElement", "value": "<hr>"},

    { "name": "SETTING_ALARM3ENABLE", "type": "ACCheckbox", "value": "1", "label": "Alarm 3 Enable", "checked": "false", "global": "true"},
    { "name": "SETTING_ALARM3SUN", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3MON", "type": "ACCheckbox", "value": "1", "label": "M", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3TUE", "type": "ACCheckbox", "value": "1", "label": "T", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3WED", "type": "ACCheckbox", "value": "1", "label": "W", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3THU", "type": "ACCheckbox", "value": "1", "label": "Th", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3FRI", "type": "ACCheckbox", "value": "1", "label": "F", "checked": "false", "posterior": "none"},
    { "name": "SETTING_ALARM3SAT", "type": "ACCheckbox", "value": "1", "label": "S", "checked": "false"},
    { "name": "SETTING_ALARM3HOUR", "type": "ACSelect", "option":  [ "00 / 12 AM", "01 / 1 AM", "02 / 2 AM", "03 / 3 AM", "04 / 4 AM", "05 / 5 AM", "06 / 6 AM", "07 / 7 AM", "08 / 8 AM", "09 / 9 AM", "10 / 10 AM", "11 / 11 AM", "12 / 12 PM", "13 / 1 PM", "14 / 2 PM", "15 / 3 PM", "16 / 4 PM", "17 / 5 PM", "18 / 6 PM", "19 / 7 PM", "20 / 8 PM", "21 / 9 PM", "22 / 10 PM", "23 / 11 PM" ], "label": "Time", "posterior": "none"},
    { "name": "SETTING_ALARM3MIN", "type": "ACSelect", "option":  [ "00", "05", "10", "15", "20", "25", "30", "35", "40", "45", "50", "55" ], "label": "", "posterior": "none"},
    { "name": "SETTING_ALARM3DUR", "type": "ACSelect", "option":  [ "Until canceled", "2 seconds", "5 seconds", "10 seconds", "30 seconds", "1 minute", "2 minutes", "3 minutes","4 minutes", "5 minutes", "6 minutes", "7 minutes", "8 minutes", "9 minutes", "10 minutes" ], "label": "Duration"},
    { "name": "newline3", "type": "ACElement", "value": "<hr>"},

    { "name": "add", "type": "ACSubmit", "value": "Submit", "uri": "/alarm"}
  ]
}
)";

**Comment by Hieromon** [comment]
@ghmpi Chopped code blocks like this are very hard to see. I can't reproduce the problem using this code.

You're probably out of memory. Enable AC_DEBUG and PB_DEBUG for trace logging. If you see ArduinoJson deserialization errors or PageBuilder errors, the cause is there.

Also, AutoConnect's JSON buffer size is defined in [`AUTOCONNECT_JSONBUFFER_SIZE`](https://github.com/Hieromon/AutoConnect/blob/f21f57d4b909a4540bea06655aef9b88e5d1b992/src/AutoConnectDefs.h#L291). Use the [ArduinoJson Assistant](https://arduinojson.org/v6/assistant/#/step1) to calculate the buffer size needed to deserialize your JSON description.

So, what's the point of your experiment? It's not clear what the point of the problem is. Move to Discussions.

**Comment by ghmpi** [comment]
Code upload 

[working-and-not-working-autoconnect.txt](https://github.com/Hieromon/AutoConnect/files/11180487/working-and-not-working-autoconnect.txt)

**Comment by ghmpi** [comment]
@Hieromon 

Thank you for your reply.

The posting interface is breaking up the pasted code, even when I encode it in code clocks.. I have uploaded it as a text file.

I did enable the debug before.. and the only thing I could obviously see was..
[AC] /alarm not registered, handler ignored

But your memory and deserialization comments got me to go back to it and I found this where the 
[AC] Deserialize:NoMemory

I will look further into the buffer size.

This is not an experiment. I'm trying to get this code working on a clock. My workaround so far has been to trim down the minute values to 5 minute increments (as seen in the sample code), so it's OK for now. But eventually I'm sure I'll need something similar and someone else will run into this scenario where they are hitting these limits as well. 

Thank you again.

Regards,
-Moses

**Comment by ghmpi** [comment, marked as answer]
@Hieromon 

I got it working! It was indeed running out of memory.

Adding the following to AutoConnectDefs.h fixed it!

#define AUTOCONNECT_JSONDOCUMENT_SIZE   (16 * 1024)

The devkitc board was working because AUTOCONNECT_JSONPSRAM_SIZE is defined larger then the regular AUTOCONNECT_JSONDOCUMENT_SIZE. (8 * 1024) on a regular board vs (16 * 1024) on boards with PSRAM.

A suggestion.. if you can have the "[AC] Deserialize:NoMemory" text/error appear on the page that the error occurred..  /alarm in my case.. instead of just dropping the page, it would help the next poor soul a few hours of troubleshooting! Not sure how easy it would be to do.

Thank you again @Hieromon , you have a neat project.

Regards,
-Moses

**Comment by Hieromon** [comment]
@ghmpi The procedure to take if you cannot see the custom web page is explained in the [FAQ](https://hieromon.github.io/AutoConnect/faq.html#i-cannot-see-the-custom-web-page) in the Documentation. It states that the first step is to validate the JSON syntax using ArduinoJson assistant. Then you enable AC_DEBUG and get the AC_DEBUG log trace.  It is not so difficult to discover the *NoMemory* error log if you track the trace. At first, you should follow the procedure, which is the closest path to the cause. It will lead you to the *NoMemory* error and will reduce the time it takes to get to the [solution](https://hieromon.github.io/AutoConnect/acjson.html#adjust-the-json-document-buffer-size) beyond that.

And this procedure is related to your following mention:

> if you can have the "[AC] Deserialize:NoMemory" text/error appear on the page that the error occurred

Unfortunately we cannot make it happen. 
ArduinoJson will not generate a JSON object if JSON parsing fails. Such cases may be due to JSON syntax errors or insufficient memory. In other words, AutoConnect has no way of knowing the existence of the `/alarm` custom web page. It is not possible to return any response HTML page to accesses to `/alarm` whose existence is unrecognized. This results in a 404 error. The only effective means is the AC_DEBUG trace.

Nevertheless, if you still want to detect and export JSON errors in the web page response, your own sketch implementation, knowing the existence of /alarm, will make this possible.

```cpp
AutoConnect portal;

void handleNotFound() {
  WebServer& server = portal.host();
  if (server.uri().endsWith("/alarm")) {
    server.send("200", "text/plain", "/alarm not loaded, check JSON error");
  }
  else {
    server.send(404);
  }
}

portal.onNotFound(handleNotFound);
```

A description of the onNotFound handler for AutoConnect can be found [here](https://hieromon.github.io/AutoConnect/adnetwork.html#404-handler).

In this case, your report allowed us to post on this topic. In Issues, similar discussions tend to hide in the background after close. In Discussions, we will have the opportunity to see them. This will provide guidance to other users to help them resolve their issues Thank you for your contribution, it is much appreciated.

**Comment by ghmpi** [comment]
@Hieromon 

Thank you for your detailed answer! I'm slowly getting into the details of your project.

If you are curious my retro clock project is using your autoconnect libraries for wifi and config pages. Here is a screenshot of the alarm page:
http://www.neonixie.com/IV-9-6D-RR/neonixie-clock-alarm-page-preview-04-09-2023.jpg

And the project in general is here: http://www.neonixie.com/IV-9-6D-RR/

I appreciate your contribution with this library!

Regards,
-Moses
